Algorithm.Framework/Portfolio/MeanVarianceOptimizationPortfolioConstruction.py

```diff
@@ -0,0 +1,138 @@
+# Lean Algorithmic Trading Engine v2.0. Copyright 2014 QuantConnect Corporation.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from clr import AddReference
+AddReference("System")
+AddReference("QuantConnect.Algorithm")
+AddReference("QuantConnect.Common")
+AddReference("QuantConnect.Indicators")
+
+from System import *
+from QuantConnect import *
+from QuantConnect.Util import PythonUtil
+from QuantConnect.Algorithm import *
+from QuantConnect.Algorithm.Framework import *
+from QuantConnect.Algorithm.Framework.Alphas import *
+from QuantConnect.Data.UniverseSelection import *
+from datetime import timedelta
+import numpy as np
+import pandas as pd
+from scipy.optimize import minimize
+
+### <summary>
+### Provides an implementation of Mean-Variance portfolio optimization based on modern portfolio theory.
+### The target percent holdings of each security is calculated by maximizing the Sharpe ratio. 
+### The interval of weights in optimization method can be changed based on the long-short algorithm.
+### The default model uses the last three months daily price to calculate the optimal weight 
+### with the weight range from -1 to 1  
+### </summary>
+class MeanVarianceOptimizationPortfolioConstructionModel:
+    def __init__(self, resolution=Resolution.Daily, interval=63, min_weight=-1, max_weight=1):
+        """ 
+        Initialize the model
+        Args:
+            resolution:The resolution of the history price 
+            interval(int): The time interval of history price to calculate the weight 
+            min_weight(float): The lower bounds on portfolio weights
+            max_weight(float): The upper bounds on portfolio weights
+        """
+        self.securities = [ ]
+        self.resolution = resolution
+        self.interval = interval
+        self.min_weight = min_weight
+        self.max_weight = max_weight
+
+    def CreateTargets(self, algorithm, insights):
+        """ 
+        Create portfolio targets from the specified insights
+        Args:
+            algorithm:The algorithm instance
+            insights: The insights to create portoflio targets from
+        Returns: 
+            An enumerable of portoflio targets to be sent to the execution model
+        """
+        
+        symbols = [insight.Symbol for insight in insights]
+        # request the daily history price of last year
+        hist = algorithm.History(symbols, self.interval, self.resolution)
```

Comment:
Instead of making a Historical data request on every `CreateTargets` call, you could make just one in `OnSecuritiesChanged`. You would use the values to update a `SimpleMovingAverage` and `StandardDeviation` indicators for each symbol.
You would have a dictionary keyed by `Symbol` with `SymbolData` (see `MacdAlphaModel` for example) will all the variables you need to compute the weights.